### PR TITLE
change folder uid for pathId

### PR DESCRIPTION
**Pull request by petersg83**
- Use `integer` for this `uid` instead of `string`
- Rename `uid` to `pathId`
- Added indexes on some columns
- Add `max` function to the `queryBuilder`

### CODE REVIEW

File: examples/getstarted/src/api/country/content-types/country/schema.json [hunk 1]
```diff
@@ -38,23 +38,6 @@
...
-      "type": "string",
-      "maxLength": 3
-    },
-    "time": {
-      "pluginOptions": {
-        "i18n": {
-          "localized": true
-        }
-      },
-      "type": "time"
```
Comment (gu-stav) [inline examples/getstarted/src/api/country/content-types/country/schema.json:57]: I guess that change is unrelated?

File: packages/core/upload/package.json [hunk 2]
```diff
@@ -41,8 +41,7 @@
     "react-redux": "7.2.3",
     "react-router": "^5.2.0",
     "react-router-dom": "5.2.0",
-    "sharp": "0.30.4",
-    "uuid": "8.3.2"
```
Comment (gu-stav) [inline packages/core/upload/package.json:45]: Why isn't that change reflected in the yarn lockfile? 🤔 I should, no?

File: packages/core/upload/server/services/folder.js [hunk 3]
```diff
@@ -1,32 +1,33 @@
...
-const generateUID = () => uuid();
+const setPathIdAndPath = async folder => {
+  const { pathId: maxPathId } = await strapi.db
+    .queryBuilder(FOLDER_MODEL_UID)
+    .max('pathId')
+    .execute();
 
-const setPathAndUID = async folder => {
-  const uid = generateUID();
+  const pathId = maxPathId + 1;
```
Comment (gu-stav) [inline packages/core/upload/server/services/folder.js:14]: I'm wondering if this could lead to concurrency problems, if two users create a folder at the same time? Before the uid was always unique, where as now, it is dependent on the previous db state.

Maybe that's just the tradeoff of the approach ...

File: packages/core/upload/tests/admin/folder.test.e2e.js [hunk 4]
```diff
@@ -15,13 +15,8 @@ let data = {
...
 };
 
-const uuidRegex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
-const rootPathRegex = /^\/[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
-const getFolderPathRegex = uid =>
-  new RegExp(
-    '^/' + uid + '/[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$',
-    'i'
-  );
+const rootPathRegex = /^\/[0-9]*$/i;
```
Comment (gu-stav) [inline packages/core/upload/tests/admin/folder.test.e2e.js:18]: That's a nice simplification :D

File: packages/core/upload/tests/content-api/graphql-upload-automatic-folder.test.e2e.js [hunk 5]
```diff
@@ -166,11 +166,11 @@ describe('Uploads folder (GraphQL)', () => {
       expect(file).toMatchObject({
         folder: {
           name: 'API Uploads (1)',
-          uid: expect.anything(),
+          pathId: expect.anything(),
```
Comment (gu-stav) [inline packages/core/upload/tests/content-api/graphql-upload-automatic-folder.test.e2e.js:169]:
```suggestion
          pathId: expect.any(Number),
```

Same for a couple of instances below.

**Review by gu-stav — COMMENTED** [review]
Overall it looks really good to me. Thanks for working on it. The only thing that bothers me a bit is that a newly created folder is now based on the previous database state, which could lead to problems when the path is built up, but I don't have an idea how to solve it, so I guess it is just a trade-off of the approach.

I'd like to have Alex's opinion before merging though :)

File: packages/core/database/lib/query/query-builder.js [hunk 6]
```diff
@@ -77,6 +78,14 @@ const createQueryBuilder = (uid, db) => {
       return this;
     },
 
+    max(column) {
+      state.type = 'max';
+      state.max = column;
```
Comment (alexandrebodin) [inline packages/core/database/lib/query/query-builder.js:83]: shouldn't we put that in the state.select so it actually uses the logic of prefixing etc that is already implemented ?

**Review by alexandrebodin — COMMENTED** [review]
Just a doubt on the db approach but the rest looks good Not much to add to gustav's comments. 

Let's document the tradeoffs in an ADR :)

File: examples/getstarted/src/api/country/content-types/country/schema.json (hunk 1 above)
Comment (petersg83) [reply examples/getstarted/src/api/country/content-types/country/schema.json:57]: Yes, unrelated :p Those fields were added by mistake a while ago and I would prefer this CT to stay simple/small

File: packages/core/upload/server/services/folder.js (hunk 3 above)
Comment (petersg83) [reply packages/core/upload/server/services/folder.js:14]: I added a unique index in the database for this column, so it won't be possible to have duplicates. If 2 requests are done at the same time, one will fail then. Regarding how we create folders we should keep database consistency.

File: packages/core/database/lib/query/query-builder.js (hunk 6 above)
Comment (petersg83) [reply packages/core/database/lib/query/query-builder.js:83]: I'm not sure to understand.
How do you use select with `qb.max(helpers.toColumnName(meta, state.max), { as: state.max });` ?

Comment (alexandrebodin) [reply packages/core/database/lib/query/query-builder.js:83]: I mean instead of doing state.max = col you could actually use state.select = col
 as it is in reality a normal select behind the hood and should follow the same rules around prefixes toCOlumnName and such
 
 And do qb.max(state.select)

**Review by gu-stav — APPROVED** [review]
✨
